src/main/java/com/google/codeu/servlets/MessageServlet.java

```diff
@@ -42,6 +42,12 @@ public void init() {
     datastore = new Datastore();
   }
 
+  String replaceImage(String source) {
+    String regex = "(?:\\[(.+)\\])?(https?:([/|.|\\w|\\s|-|%])*\\.(?:jpg|gif|png))";
```

Comment:
This regex doesn't match image URL that contains letter '-' (hyphen). Example: https://sta.codeforces.com/s/21920/images/codeforces-logo-with-telegram.png (Codeforces logo).

Also, what does the `(?:\\[(.+)\\])` do?

Reply:
Yeah I didn't escape the hyphen character so it didn't work. It has been fixed with commit d960d1a.

With `(?:\\[(.+)\\])` I want to capture the image title, so users can write [image-title]URL.

Reply:
I recommend you to change it to [image-title](imageURL) as in Markdown, so it would be a bit easier to read for the users.

Other than that, it works fine now.

Reply:
Yeah I just think that we are not expecting users to know that syntax of Markdown, so I will keep it simple for now.